mrp_production_real_cost/models/stock_move.py

```diff
@@ -37,30 +38,58 @@ def action_done(self):
+            current_available = 0.0
+        current_value = current_available * current_price
+        if (current_available + moved) <= 0:
+            return 0.0
+        return (
+            (current_value + cost) /
+            (current_available + moved)
+        )
+
     @api.multi
     def product_price_update_production_done(self):
         records = self.filtered(
             lambda x: (x.production_id and
                        x.product_id.cost_method == 'average'))
+        products = {}
+        templates = {}
         for move in records:
-            prod_total_cost = move.production_id.real_cost
             product = move.product_id
-            product_avail = product.qty_available
-            amount_unit = product.standard_price
-            tmpl_available = product.product_tmpl_id.qty_available
-            tmpl_price = product.product_tmpl_id.standard_price
+            tmpl = product.product_tmpl_id
+            product_data = products.get(product.id, False) or {}
+            if not product_data:
+                product_data['product'] = product
+                product_data['available'] = product.qty_available
+                product_data['price'] = product.standard_price
+                product_data['moved'] = 0.0
+                product_data['cost'] = move.production_id.real_cost
+            tmpl_data = templates.get(tmpl.id, False) or {}
+            if not tmpl_data:
+                tmpl_data['template'] = tmpl
+                tmpl_data['available'] = tmpl.qty_available
+                tmpl_data['price'] = tmpl.standard_price
+                tmpl_data['moved'] = 0.0
+                tmpl_data['cost'] = move.production_id.real_cost
             if move.state == 'done':
-                product_avail -= move.product_qty
-                tmpl_available -= move.product_qty
-            new_product_price = (
-                (amount_unit * product_avail + prod_total_cost) /
-                ((product_avail >= 0.0 and product_avail or 0.0) +
-                 move.product_qty))
-            new_tmpl_price = ((tmpl_price * tmpl_available + prod_total_cost) /
-                              ((tmpl_available > 0.0 and tmpl_available or
-                                0.0) + move.product_qty))
-            product.sudo().standard_price = new_product_price
-            product.sudo().product_tmpl_id.standard_price = new_tmpl_price
+                product_data['available'] -= move.product_qty
+                product_data['moved'] += move.product_qty
+                tmpl_data['available'] -= move.product_qty
+                tmpl_data['moved'] += move.product_qty
+            products[product.id] = product_data
+            templates[tmpl.id] = tmpl_data
+        for product_id, product_data in products.iteritems():
+            new_price = self._new_average_price(product_data),
+            product_data['product'].sudo().standard_price = new_price
+        for tmpl_id, tmpl_data in templates.iteritems():
```

Comment:
And that's the key: if you don't install it, variant cost = template cost. Writing to product is the same as writing to template. If you have it installed, there are 2 cases:

* There are more than 1 variants: the cost price is only written to the variant.
* There are only 1 variant: the cost price in written on the variant and then written upwards to the template.

In any case, you only need to write to `product.product`. Inheritance and the other module do the rest.